akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala

```diff
@@ -208,14 +208,16 @@ class HubSpec extends StreamSpec {
       draining.drainAndComplete()
 
       downstream.request(10)
-      downstream.expectNextN(10).sorted should ===(1 to 10)
+      val firstBatch = downstream.expectNextN(10)
 
       val upstream = TestPublisher.probe[Int]()
       Source.fromPublisher(upstream).runWith(sink)
       upstream.expectCancellation()
```

Comment:
Did you consider my suggestion to make the upstreams also probes so that you can see demand from the hub to know that it is running, and potentially stall the drainAndComplete by not sending anything after a pull in one of the upstreams while using the other one to observe cancellation?

Reply:
Yes I considered it, however, if I understood it correctly, I do not believe it solves the issue.

The test is meant to verify that once the `MergeHub` is set to draining all new producers are canceled immediately.

Using the probes as you mentioned would indeed allow me to verify that the `MergeHub` has been started and keep it stalled while I attach a new producer. This however is already covered by the current test by splitting the consumption of the stream into 2 batches: after the first read of 10 elements we know that the `MergeHub` has been started and it is stalled since it still needs to consume the remaining 10 elements.

The issue comes when I attach the new producer: knowing that the `MergeHub` is already started is not a 100% guarantee that it also has processed the draining `AsyncCallback`.

So at this point when the new producers gets materialized there are 2 possibilities:
- the draining callback has been processed: the new producer get cancelled causing the `expectCancellation` succeeds
- the draining callback has not been processed: the new producer is attached to the `MergeHub`  causing the `expectCancellation` to timeout (since the the `MergeHub` is stalled by the remaining 10 elements)

I do not see any way to use the probes to infer the internal state of the `MergeHub` draining flag so that I can defer the materialization of the new producer to a point in time where it is guaranteed to be in the desired state (i.e. draining has been started).

Still, I do believe that such a scenario is a pretty unlucky one as the `AsyncCallback` would have to be processed after:
1. the consumption of 10 elements
2. the materialization of the new producer

Reply:
Ah, I see, I misremembered that the hub cancels the upstream after `drainAndComplete` but it waits for them to complete, so there is no way to observe that, I agree, it sounds unlikely that this particular race will happen so let's go with the current fix (and we'll revisit in case it turns out it does in fact ever happen in our nightly CI-tests) 👍